Define a sentence as a word* without a NEW_LINE at the end.

benignbala opened the pull request:
Using 'phrase' for the header causes errors because the phrase itself
has a NEW_LINE and the header was a phrase followed by a NEW_LINE. So,
header became word\* NEW_LINE NEW_LINE which is incorrect. Now, header
is sentence NEW_LINE, which is then word\* NEW_LINE.

benignbala committed:
Define a sentence as a word* without a NEW_LINE at the end.

Using 'phrase' for the header causes errors because the phrase itself
has a NEW_LINE and the header was a phrase followed by a NEW_LINE. So,
header became word* NEW_LINE NEW_LINE which is incorrect. Now, header
is sentence NEW_LINE, which is then word* NEW_LINE.

abelsromero commented:
I don't see any problem with it.

benignbala commented:
Without this PR, it doesn't parse the header for simple_example.adoc. Thanks

PS: I did reply earlier through email, it somehow didn't reach here. :(

abelsromero commented:
> Without this PR, it doesn't parse the header for simple_example.adoc. Thanks

I was that, did you have a change to look into other issues? I really have problems understanding why words ending with punctuation are not correctly processed?

benignbala commented:
@abelsromero Sure, I'll have a go at that :). Thanks

benignbala committed:
Punctuation works for 'period'

The parsing of punctuation now works for the 'period'. Need to get others
working.

benignbala committed:
Fixes for punctuation and attributes.

A few more punctations work now. And initial support for predefined
attributes added. At the moment, looking at just the English language.
Will have to extend to other languages once this works fine.

benignbala commented:
A collection of changes that seem to work now :) Punctuations work, at least to some extent. Thanks

abelsromero commented on src/main/antlr/Asciidoc.g4:
```diff
@@ -48,7 +49,7 @@ word            : CHARS PUNCTUATION? ;
+PUNCTUATION     : '.' ;//[.:,;] ;
 //word            : ~[\n\t ]+ ;
-ANYCHAR         : ~[\n\t ]+ ; // PH
+ANY_CHAR         : ~[\n\t ] ; // PH
```
Are you sure about this? Without the `+` most words fail to parse for me.

benignbala replied:
This is parsing most of the doc for me - The only thing that is failing now is a punctuation following a word without a space. I'm looking at some Antlr docs to see how we can handle this.

abelsromero replied:
Odd, with `ANY_CHAR: ~[\n\t ] ;` it does not parse the body for me
![1_original](https://cloud.githubusercontent.com/assets/5781153/22325482/8fc5ac2e-e3ae-11e6-8d18-9e36cb6a64f4.png)

But with the`+`, it does
![2_with](https://cloud.githubusercontent.com/assets/5781153/22325484/9368cc4e-e3ae-11e6-969b-7407fdc66b0d.png)

abelsromero commented:
Some improvements really! I see that headers and author sections are parsed now, the `simple-document.adoc` is almost parsed :smiley: However, the punctuation issues persist, words ending in punctuation signs fail.
![dot_issue](https://cloud.githubusercontent.com/assets/5781153/22324047/12af1330-e3a7-11e6-9f21-8b8be1922ded.PNG)
Note: by now, ignore the issue in `haven't`

I think it is because the tokenizer symbol `WS  : [ \t\r]+ -> skip ;`, this is just wrong `word  : CHARS PUNCTUATION? | CHARS SPLPUNCTUATION CHARS;` It expects a space between the letters and the trailing punctuation. I fear we need to define a special symbol that concatenates letters and sings.

benignbala commented:
@abelsromero Thanks :) BTW, is this word: CHARS PUNCTUATION? | CHARS SPLPUNCTUATION CHARS; is needed because of words that have a punctuation like "haven't" ,"it's" and "cross-legged" that have a punctuation in the middle of a word. And like you noticed, it seems to be parsing punctuations fine if there is a space between the word and the punctuation. Will check what's going wrong there. Thanks

![section_body](https://cloud.githubusercontent.com/assets/600098/22325362/a1b5fcf2-e3d3-11e6-8db8-c44cf4e07b19.png)
